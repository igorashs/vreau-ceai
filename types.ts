import FormidableLib from 'formidable';
import { ValidationError } from 'joi';

export type Product = {
  _id: string;
  name: string;
  category_id: string;
  src: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
};

export type ProductWithCategory = Product & {
  category_id: { name: string };
};

export type ErrorDetail = {
  message: string;
  name?: string;
};

export type ProductName = {
  name: string;
};

export type ProductNameErrorDetail = ErrorDetail & {
  name?: keyof ProductName;
};

export type ProductFields = ProductName & {
  src?: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
  category_id: string;
};

export type Category = {
  _id: string;
  name: string;
};

export type ProductErrorDetail = ErrorDetail & {
  name?: keyof ProductFields;
};

export type ApiResponse = {
  success: boolean;
  message: string;
  errors?: ErrorDetail[];
};

export type User = {
  _id: string;
  name: string;
  email: string;
  isManager: boolean;
};

export type UserEmail = {
  email: string;
};

export type UserEmailErrorDetail = ErrorDetail & {
  name?: keyof UserEmail;
};

export type UserLogin = UserEmail & { password: string };

export type UserSignup = UserEmail & {
  name: string;
  password: string;
  repeat_password: string;
};

export type UserPermissions = { isManager: boolean };

export type LabelMessage = { success: boolean; message: string };

export type UserAuth = {
  _id: string;
  name: string;
  isAdmin: boolean;
  isManager: boolean;
};

export type UserSession = {
  isAuth: boolean;
  user: UserAuth | null;
  needRefresh: boolean;
};

type dataType = 'data' | 'error' | 'field' | 'fileBegin' | 'file' | 'progress';

export declare class Formidable extends FormidableLib {
  emit(name: dataType, data: FormidableLib.EmitData | ValidationError): void;
}

export type CategoryName = {
  name: string;
};

export type CategoryNameErrorDetail = ErrorDetail & {
  name?: keyof CategoryName;
};

export type OrderSubmit = {
  tel: string;
  address: string;
};

export type OrderItem = {
  product_id: string;
  count: number;
};

export type OrderStatus = {
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
};

export type OrderStatusErrorDetail = ErrorDetail & {
  name?: keyof OrderStatus;
};

export type OrderNumber = {
  number: string;
};

export type OrderNumberErrorDetail = ErrorDetail & {
  name?: keyof OrderNumber;
};

export type OrderFields = {
  info: OrderSubmit;
  items: OrderItem[];
};

export type Order = {
  _id: string;
  user: {
    name: string;
    email: string;
  };
  number: string;
  total_price: number;
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
  items: {
    count: number;
    product: { name: string; price: number; quantity: number };
  }[];
  address: string;
  tel: string;
  orderedAt: Date;
  completedAt: Date;
};

export type UserOrders = { orders: Order[]; count: number };

export type ManagementOrders = { orders: Order[]; count: number };
