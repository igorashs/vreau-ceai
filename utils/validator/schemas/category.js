import Joi from 'joi';

export const categorySchema = Joi.object({
  name: Joi.string().trim().max(30).required().messages({
    'string.base': 'numele trebuie sa fie de tip text',
    'string.max': 'numele este prea lung',
    'string.empty': 'numele este obligatoriu',
    'any.required': 'numele este obligatoriu'
  })
});
