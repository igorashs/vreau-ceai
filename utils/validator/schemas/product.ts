import Joi from 'joi';
import { ProductFields, ProductName } from 'types';

export const productMessages = {
  name: {
    type: 'numele trebuie sa fie de tip text',
    max: 'numele este prea lung',
    required: 'numele este obligatoriu',
  },

  price: {
    type: 'prețul trebuie să fie valid',
    min: 'prețul nu poate fi negativ',
    invalid: 'valoarea introdusă este invalidă',
  },

  quantity: {
    type: 'cantitatea trebuie să fie validă',
    min: 'cantitatea nu poate fi negativă',
    invalid: 'valoarea introdusă este invalidă',
  },

  total_quantity: {
    type: 'cantitatea totală trebuie să fie validă',
    min: 'cantitatea totaă nu poate fi negativă',
    invalid: 'valoarea introdusă este invalidă',
  },

  description: {
    type: 'descrierea nu este validă',
    max: 'descrierea este prea lungă',
    required: 'descrierea este obligatorie',
  },

  category_id: {
    required: 'categoria este obligatorie',
  },
};

const name = Joi.string().trim().max(60).required().messages({
  'string.base': productMessages.name.type,
  'string.max': productMessages.name.max,
  'any.required': productMessages.name.required,
  'string.empty': productMessages.name.required,
});

export const nameSchema = Joi.object<ProductName>({ name });

export const productSchema = Joi.object<ProductFields>({
  name,

  src: Joi.any().default('placeholder.png'),

  price: Joi.number().min(0).default(0).messages({
    'number.base': productMessages.price.type,
    'number.min': productMessages.price.min,
    'number.unsafe': productMessages.price.invalid,
  }),

  quantity: Joi.number().min(0).default(0).messages({
    'number.base': productMessages.quantity.type,
    'number.min': productMessages.quantity.min,
    'number.unsafe': productMessages.quantity.invalid,
  }),

  total_quantity: Joi.number().min(0).default(0).messages({
    'number.base': productMessages.total_quantity.type,
    'number.min': productMessages.total_quantity.min,
    'number.unsafe': productMessages.total_quantity.invalid,
  }),

  description: Joi.string().trim().max(2000).required().messages({
    'string.base': productMessages.description.type,
    'string.max': productMessages.description.max,
    'string.empty': productMessages.description.required,
    'any.required': productMessages.description.required,
  }),

  recommend: Joi.boolean().default(false),

  category_id: Joi.string().required().messages({
    'any.required': productMessages.category_id.required,
  }),
});
