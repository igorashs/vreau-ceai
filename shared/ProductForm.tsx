/* eslint-disable camelcase */
import { useForm } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/dist/ie11/joi';
import { productSchema } from '@/utils/validator/schemas/product';
import { Form, FormAction } from '@/shared/Form';
import { TextField } from '@/shared/TextField';
import CheckBox from '@/shared/CheckBox';
import Button from '@/shared/Button';
import { Select } from '@/shared/Select';
import { InputFile } from '@/shared/InputFile';
import { Textarea } from '@/shared/Textarea';
import { useEffect } from 'react';

type Category = {
  _id: string;
  name: string;
};

type Product = {
  _id: string;
  name: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  category_id: string;
  recommend: boolean;
  src?: FileList;
};

type ProductInputs = Omit<Product, '_id'>;

type InputsErrors = Array<{
  message: string;
  name:
    | 'name'
    | 'price'
    | 'quantity'
    | 'total_quantity'
    | 'description'
    | 'category_id'
    | 'recommend'
    | 'src';
}>;

interface ProductFormProps {
  onProductSubmit: (data: ProductInputs) => Promise<InputsErrors>;
  categories: Array<Category>;
  product: Product;
}

export function ProductForm({
  onProductSubmit,
  categories,
  product,
}: ProductFormProps) {
  const {
    register,
    handleSubmit,
    setError,
    reset,
    watch,
    formState: { errors },
  } = useForm<ProductInputs>({
    mode: 'onChange',
    resolver: joiResolver(productSchema),
  });

  const watchSrc = watch('src');

  useEffect(() => {
    reset({
      name: product?.name ?? '',
      price: product?.price ?? 0,
      quantity: product?.quantity ?? 0,
      total_quantity: product?.total_quantity ?? 0,
      description: product?.description ?? '',
      category_id: product?.category_id,
      recommend: product?.recommend ?? false,
      src: undefined,
    });
  }, [product, categories]);

  const onSubmit = async (data: ProductInputs) => {
    const submitErrors = await onProductSubmit(data);

    if (submitErrors) {
      submitErrors.forEach((error) => {
        const { message, name } = error;
        setError(name, { message });
      });
    }
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <TextField
        id={product && `name_${product._id}`}
        name="name"
        label="nume produs"
        error={errors?.name?.message}
        passRef={register}
        type="text"
      />
      <TextField
        id={product && `price_${product._id}`}
        name="price"
        label="preț (lei)"
        error={errors?.price?.message}
        passRef={register}
        type="number"
      />
      <TextField
        id={product && `quantity_${product._id}`}
        name="quantity"
        label="preț per cantitate (g)"
        error={errors?.quantity?.message}
        passRef={register}
        type="number"
      />
      <TextField
        id={product && `total_quantity_${product._id}`}
        name="total_quantity"
        label="cantitatea totală (g)"
        error={errors?.total_quantity?.message}
        passRef={register}
        type="number"
      />
      <Textarea
        id={product && `description_${product._id}`}
        name="description"
        label="descriere"
        passRef={register}
        error={errors?.description?.message}
        rows={7}
      />
      <Select
        id={product && `category_id_${product._id}`}
        name="category_id"
        label="categorie"
        passRef={register}
        error={errors?.category_id?.message}
      >
        {categories ? (
          categories.map((c) => (
            <option key={c._id} value={c._id}>
              {c.name}
            </option>
          ))
        ) : (
          <option>no category</option>
        )}
      </Select>
      <InputFile
        id={product && `src_${product._id}`}
        name="src"
        label={`(max 1MB) imagine ${(watchSrc && watchSrc[0]?.name) || ''}`}
        passRef={register}
        error={errors?.src?.message}
        accept=".png, .jpg, .jpeg"
      >
        Încarcă imagine
      </InputFile>
      <FormAction justify="space-between">
        <CheckBox
          id={product && `recommend_${product._id}`}
          name="recommend"
          label="recomandată"
          passRef={register}
        />
        <Button>salvează</Button>
      </FormAction>
    </Form>
  );
}
