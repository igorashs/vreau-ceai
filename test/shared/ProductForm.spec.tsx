import {
  cleanup,
  render,
  screen,
  userEvent,
  waitFor,
} from '@/utils/test-utils';
import ProductForm from '@/shared/ProductForm';
import { Category, ProductFields, Product, ProductErrorDetail } from 'types';
import { productMessages } from '@/utils/validator/schemas/product';

const onProductSubmitMock = jest.fn(
  (): Promise<ProductErrorDetail[] | undefined> => Promise.resolve(undefined),
);

const categories: Category[] = [
  { _id: 'c1', name: 'green' },
  { _id: 'c2', name: 'fruit' },
];

describe('ProductForm', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ProductForm
        onProductSubmit={onProductSubmitMock}
        categories={categories}
      />,
    );

    expect(screen.getByRole('textbox', { name: /nume produs/i })).toHaveValue(
      '',
    );

    expect(
      screen.getByRole('spinbutton', { name: /preț \(lei\)/i }),
    ).toHaveValue(0);

    expect(
      screen.getByRole('spinbutton', { name: /preț per cantitate \(g\)/i }),
    ).toHaveValue(0);

    expect(
      screen.getByRole('spinbutton', { name: /cantitatea totală \(g\)/i }),
    ).toHaveValue(0);

    expect(screen.getByRole('textbox', { name: /descriere/i })).toHaveValue('');
    expect(screen.getByRole('combobox', { name: /categorie/i })).toHaveValue(
      categories[0]._id,
    );

    expect(screen.getAllByRole('option')).toHaveLength(categories.length);
    expect(screen.getByLabelText(/Încarcă imagine/i)).toHaveValue('');
    expect(
      screen.getByRole('checkbox', { name: /recomandată/i }),
    ).not.toBeChecked();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('has optional product prop for initial values', () => {
    const product: Product = {
      _id: 'p1',
      category_id: categories[0]._id,
      description: 'sweet tea',
      name: 'Green Tea',
      price: 69,
      quantity: 50,
      recommend: true,
      src: '',
      total_quantity: 34500,
    };

    render(
      <ProductForm
        onProductSubmit={onProductSubmitMock}
        categories={categories}
        product={product}
      />,
    );

    expect(screen.getByRole('textbox', { name: /nume produs/i })).toHaveValue(
      product.name,
    );

    expect(
      screen.getByRole('spinbutton', { name: /preț \(lei\)/i }),
    ).toHaveValue(product.price);

    expect(
      screen.getByRole('spinbutton', { name: /preț per cantitate \(g\)/i }),
    ).toHaveValue(product.quantity);

    expect(
      screen.getByRole('spinbutton', { name: /cantitatea totală \(g\)/i }),
    ).toHaveValue(product.total_quantity);

    expect(screen.getByRole('textbox', { name: /descriere/i })).toHaveValue(
      product.description,
    );
    expect(screen.getByRole('combobox', { name: /categorie/i })).toHaveValue(
      product.category_id,
    );
    expect(screen.getByLabelText(/Încarcă imagine/i)).toHaveValue(product.src);
    expect(
      screen.getByRole('checkbox', { name: /recomandată/i }),
    ).toBeChecked();
  });

  describe('fields validation', () => {
    beforeEach(() =>
      render(
        <ProductForm
          onProductSubmit={onProductSubmitMock}
          categories={categories}
        />,
      ),
    );

    describe('name field', () => {
      it('should display required error', async () => {
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.name.required),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display max length error', async () => {
        userEvent.type(
          screen.getByRole('textbox', { name: /nume produs/i }),
          'more than 60 chars|111111111111111111111111111111111111111111',
        );
        userEvent.click(screen.getByRole('button'));

        expect(await screen.findByText(productMessages.name.max)).toBeVisible();
        expect(onProductSubmitMock).not.toBeCalled();
      });
    });

    describe('price field', () => {
      it('should display negative number error', async () => {
        userEvent.type(
          screen.getByRole('spinbutton', { name: /preț \(lei\)/i }),
          '{backspace}-1',
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.price.min),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display invalid error', async () => {
        const priceInput = screen.getByRole('spinbutton', {
          name: /preț \(lei\)/i,
        });

        // big number
        userEvent.type(priceInput, '9007199254740992');

        expect(
          await screen.findByText(productMessages.price.invalid),
        ).toBeVisible();

        // empty field
        userEvent.type(priceInput, '{backspace}', { initialSelectionEnd: -1 });
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.price.invalid),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display type error', async () => {
        userEvent.type(
          screen.getByRole('spinbutton', { name: /preț \(lei\)/i }),
          'not a number',
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.price.type),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });
    });

    describe('quantity field', () => {
      it('should display negative number error', async () => {
        userEvent.type(
          screen.getByRole('spinbutton', { name: /preț per cantitate \(g\)/i }),
          '{backspace}-1',
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.quantity.min),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display invalid error', async () => {
        const quantityInput = screen.getByRole('spinbutton', {
          name: /preț per cantitate \(g\)/i,
        });

        // big number
        userEvent.type(quantityInput, '9007199254740992');

        expect(
          await screen.findByText(productMessages.quantity.invalid),
        ).toBeVisible();

        // empty field
        userEvent.type(quantityInput, '{backspace}', {
          initialSelectionEnd: -1,
        });
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.quantity.invalid),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display type error', async () => {
        userEvent.type(
          screen.getByRole('spinbutton', {
            name: /preț per cantitate \(g\)/i,
          }),
          'not a number',
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.quantity.type),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });
    });

    describe('total_quantity field', () => {
      it('should display negative number error', async () => {
        userEvent.type(
          screen.getByRole('spinbutton', { name: /cantitatea totală \(g\)/i }),
          '{backspace}-1',
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.total_quantity.min),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display invalid error', async () => {
        const totalQuantityInput = screen.getByRole('spinbutton', {
          name: /cantitatea totală \(g\)/i,
        });

        // big number
        userEvent.type(totalQuantityInput, '9007199254740992');

        expect(
          await screen.findByText(productMessages.total_quantity.invalid),
        ).toBeVisible();

        // empty field
        userEvent.type(totalQuantityInput, '{backspace}', {
          initialSelectionEnd: -1,
        });
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.total_quantity.invalid),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display type error', async () => {
        userEvent.type(
          screen.getByRole('spinbutton', { name: /cantitatea totală \(g\)/i }),
          'not a number',
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.total_quantity.type),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });
    });

    describe('description field', () => {
      it('should display required error', async () => {
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.description.required),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });

      it('should display max length error', async () => {
        const hugeDescription =
          'More than 2000 chars|sit amet, consectetur adipiscing elit. Quisque porta congue nisl sed imperdiet. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Duis in erat velit. Nulla magna velit, auctor sed ornare ac, laoreet tristique dolor. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Morbi erat ex, efficitur sit amet vulputate nec, bibendum ut tellus. Curabitur ac interdum metus. Duis nec elit eget nisi ultrices eleifend in eu odio.Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque porta congue nisl sed imperdiet. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Duis in erat velit. Nulla magna velit, auctor sed ornare ac, laoreet tristique dolor. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Morbi erat ex, efficitur sit amet vulputate nec, bibendum ut tellus. Curabitur ac interdum metus. Duis nec elit eget nisi ultrices eleifend in eu odio.Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque porta congue nisl sed imperdiet. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Duis in erat velit. Nulla magna velit, auctor sed ornare ac, laoreet tristique dolor. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Morbi erat ex, efficitur sit amet vulputate nec, bibendum ut tellus. Curabitur ac interdum metus. Duis nec elit eget';
        userEvent.type(
          screen.getByRole('textbox', { name: /descriere/i }),
          hugeDescription,
        );
        userEvent.click(screen.getByRole('button'));

        expect(
          await screen.findByText(productMessages.description.max),
        ).toBeVisible();

        expect(onProductSubmitMock).not.toBeCalled();
      });
    });

    describe('category_id field', () => {
      it('should display no category', async () => {
        cleanup();

        render(
          <ProductForm onProductSubmit={onProductSubmitMock} categories={[]} />,
        );

        userEvent.click(screen.getByRole('button'));

        expect(await screen.findByText('no category')).toBeVisible();
        expect(onProductSubmitMock).not.toBeCalled();
      });
    });
  });

  describe('success submision', () => {
    it('should reset the form', async () => {
      const product: ProductFields & { src: string } = {
        category_id: categories[1]._id,
        description: 'sweet tea',
        name: 'Green Tea',
        price: 69,
        quantity: 50,
        recommend: true,
        src: 'tea.png',
        total_quantity: 34500,
      };

      render(
        <ProductForm
          onProductSubmit={onProductSubmitMock}
          categories={categories}
        />,
      );

      const nameInput = screen.getByRole('textbox', { name: /nume produs/i });
      const priceInput = screen.getByRole('spinbutton', {
        name: /preț \(lei\)/i,
      });
      const qunatityInput = screen.getByRole('spinbutton', {
        name: /preț per cantitate \(g\)/i,
      });
      const totalQuantityInput = screen.getByRole('spinbutton', {
        name: /cantitatea totală \(g\)/i,
      });
      const descriptionInput = screen.getByRole('textbox', {
        name: /descriere/i,
      });
      const categorySelect = screen.getByRole('combobox', {
        name: /categorie/i,
      });
      const fileInput = screen.getByLabelText(/Încarcă imagine/i);
      const recommendationInput = screen.getByRole('checkbox', {
        name: /recomandată/i,
      });

      userEvent.type(nameInput, product.name);
      userEvent.type(priceInput, `{backspace}${product.price}`);
      userEvent.type(qunatityInput, `{backspace}${product.quantity}`);
      userEvent.type(
        totalQuantityInput,
        `{backspace}${product.total_quantity}`,
      );
      userEvent.type(descriptionInput, product.description);
      userEvent.selectOptions(categorySelect, product.category_id);
      userEvent.upload(
        fileInput,
        new File(['tea img'], product.src, { type: 'image/png' }),
      );
      userEvent.click(recommendationInput);
      userEvent.click(screen.getByRole('button'));

      await waitFor(() => expect(onProductSubmitMock).toBeCalled());

      expect(nameInput).toHaveValue('');
      expect(priceInput).toHaveValue(0);
      expect(qunatityInput).toHaveValue(0);
      expect(totalQuantityInput).toHaveValue(0);
      expect(descriptionInput).toHaveValue('');
      expect(recommendationInput).not.toBeChecked();
      expect(categorySelect).toHaveValue(categories[0]._id);
      expect(fileInput).toHaveValue('');
    });
  });

  describe('server validation', () => {
    it('should display server validation errors', async () => {
      const product: Product = {
        _id: 'p1',
        category_id: categories[0]._id,
        description: 'sweet tea',
        name: 'Green Tea',
        price: 69,
        quantity: 50,
        recommend: true,
        src: '',
        total_quantity: 34500,
      };

      // pretend user bypass client validation
      onProductSubmitMock.mockResolvedValueOnce([
        { message: productMessages.name.required, name: 'name' },
        { message: productMessages.price.invalid, name: 'price' },
        { message: productMessages.quantity.min, name: 'quantity' },
        {
          message: productMessages.total_quantity.type,
          name: 'total_quantity',
        },
        { message: productMessages.description.max, name: 'description' },
        { message: productMessages.category_id.invalid, name: 'category_id' },
        { message: productMessages.src.max, name: 'src' },
      ]);

      render(
        <ProductForm
          onProductSubmit={onProductSubmitMock}
          categories={categories}
        />,
      );

      const nameInput = screen.getByRole('textbox', { name: /nume produs/i });
      const priceInput = screen.getByRole('spinbutton', {
        name: /preț \(lei\)/i,
      });
      const qunatityInput = screen.getByRole('spinbutton', {
        name: /preț per cantitate \(g\)/i,
      });
      const totalQuantityInput = screen.getByRole('spinbutton', {
        name: /cantitatea totală \(g\)/i,
      });
      const descriptionInput = screen.getByRole('textbox', {
        name: /descriere/i,
      });
      const categorySelect = screen.getByRole('combobox', {
        name: /categorie/i,
      });
      const fileInput = screen.getByLabelText(/Încarcă imagine/i);
      const recommendationInput = screen.getByRole('checkbox', {
        name: /recomandată/i,
      });

      userEvent.type(nameInput, product.name);
      userEvent.type(priceInput, `{backspace}${product.price}`);
      userEvent.type(qunatityInput, `{backspace}${product.quantity}`);
      userEvent.type(
        totalQuantityInput,
        `{backspace}${product.total_quantity}`,
      );
      userEvent.type(descriptionInput, product.description);
      userEvent.selectOptions(categorySelect, product.category_id);
      userEvent.upload(
        fileInput,
        new File(['tea img'], product.src, { type: 'image/png' }),
      );
      userEvent.click(recommendationInput);
      userEvent.click(screen.getByRole('button'));

      await waitFor(() => expect(onProductSubmitMock).toBeCalled());

      expect(screen.getByText(productMessages.name.required)).toBeVisible();
      expect(screen.getByText(productMessages.price.invalid)).toBeVisible();
      expect(screen.getByText(productMessages.quantity.min)).toBeVisible();
      expect(
        screen.getByText(productMessages.total_quantity.type),
      ).toBeVisible();
      expect(screen.getByText(productMessages.description.max)).toBeVisible();
      expect(
        screen.getByText(productMessages.category_id.invalid),
      ).toBeVisible();
      expect(screen.getByText(productMessages.src.max)).toBeVisible();

      // form still has old values
      expect(nameInput).toHaveValue(product.name);
    });
  });
});
