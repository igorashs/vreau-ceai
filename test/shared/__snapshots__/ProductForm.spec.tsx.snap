// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
}

.c13 {
  margin-top: calc(var(--baseline) / 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c2 {
  font-weight: 500;
}

.c15 {
  font-weight: 500;
  color: var(--accent-text-dark);
  font-weight: 400;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c3 {
  width: 100%;
  background-color: var(--input);
  border: 1px solid var(--layout-dark);
  border-radius: 4px;
  padding: calc(var(--baseline) / 4) 0 calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  color: var(--text-dark);
}

.c3::-webkit-outer-spin-button,
.c3::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c3[type='number'] {
  -moz-appearance: textfield;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
}

.c16 {
  margin: 0;
  width: 16px;
  height: 16px;
  -webkit-order: -1;
  -ms-flex-order: -1;
  order: -1;
  cursor: pointer;
}

.c17 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c17:hover {
  background-color: var(--accent);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c7 {
  position: relative;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c8 {
  width: 100%;
  padding: 7px 38px 7px 14px;
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
}

.c8:hover {
  background-color: var(--accent);
}

.c8 option {
  padding: 7px 14px;
}

.c9 {
  position: absolute;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  color: var(--text-light);
  right: 7px;
  top: calc(50% - 12px);
  pointer-events: none;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c11 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c11:hover {
  background-color: var(--accent);
}

.c12 {
  display: none;
}

.c4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c5 {
  width: 100%;
  background-color: var(--input);
  border: 1px solid var(--layout-dark);
  border-radius: 4px;
  padding: calc(var(--baseline) / 4) 0 calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  color: var(--text-dark);
  resize: none;
}

<form
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="name"
    >
      nume produs
    </label>
    <input
      class="c3"
      id="name"
      name="name"
      type="text"
    />
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="price"
    >
      preț (lei)
    </label>
    <input
      class="c3"
      id="price"
      name="price"
      type="number"
    />
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="quantity"
    >
      preț per cantitate (g)
    </label>
    <input
      class="c3"
      id="quantity"
      name="quantity"
      type="number"
    />
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="total_quantity"
    >
      cantitatea totală (g)
    </label>
    <input
      class="c3"
      id="total_quantity"
      name="total_quantity"
      type="number"
    />
  </div>
  <div
    class="c4"
  >
    <label
      class="c2"
      for="description"
    >
      descriere
    </label>
    <textarea
      class="c5"
      id="description"
      name="description"
      rows="7"
    />
  </div>
  <div
    class="c6"
  >
    <label
      class="c2"
      for="category_id"
    >
      categorie
    </label>
    <div
      class="c7"
    >
      <select
        class="c8"
        id="category_id"
        name="category_id"
      >
        <option
          value="c1"
        >
          green
        </option>
        <option
          value="c2"
        >
          fruit
        </option>
      </select>
      <svg
        class="c9"
      />
    </div>
  </div>
  <div
    class="c10"
  >
    <label
      class="c2"
      for="src"
    >
      (max 1MB) imagine 
    </label>
    <label
      class="c11"
    >
      <input
        accept=".png, .jpg, .jpeg"
        class="c12"
        id="src"
        name="src"
        type="file"
      />
      Încarcă imagine
    </label>
  </div>
  <div
    class="c13"
  >
    <div
      class="c14"
    >
      <label
        class="c15"
        for="recommend"
      >
        recomandată
      </label>
      <input
        class="c16"
        id="recommend"
        name="recommend"
        type="checkbox"
      />
    </div>
    <button
      class="c17"
    >
      salvează
    </button>
  </div>
</form>
`;
